File: app/routes/mqtt.py
"""MQTT/IoT device control routes"""
from flask import Blueprint, jsonify, request, current_app
import json
import os
from app.modules.mqtt_tasmota import get_mqtt_client, init_mqtt_client, shutdown_mqtt_client

mqtt_bp = Blueprint('mqtt', __name__)

# Path to MQTT configuration file
MQTT_CONFIG_FILE = os.path.join(
    os.path.dirname(os.path.dirname(os.path.dirname(__file__))),
    'configs', 'mqtt_config.json'
)


def load_mqtt_config():
    """Load MQTT configuration from file"""
    try:
        if os.path.exists(MQTT_CONFIG_FILE):
            with open(MQTT_CONFIG_FILE, 'r') as f:
                return json.load(f)
        return None
    except Exception as e:
        current_app.logger.error(f"Error loading MQTT config: {e}")
        return None


def save_mqtt_config(config):
    """Save MQTT configuration to file"""
    try:
        with open(MQTT_CONFIG_FILE, 'w') as f:
            json.dump(config, f, indent=2)
        return True
    except Exception as e:
        current_app.logger.error(f"Error saving MQTT config: {e}")
        return False


@mqtt_bp.route('/status', methods=['GET'])
def get_mqtt_status():
    """Get MQTT connection status"""
    try:
        client = get_mqtt_client()
        config = load_mqtt_config()
        
        if client:
            return jsonify({
                'success': True,
                'connected': client.is_connected(),
                'broker': {
                    'host': client.broker_host,
                    'port': client.broker_port
                },
                'device_count': len(client.devices)
            })
        else:
            return jsonify({
                'success': True,
                'connected': False,
                'broker': config.get('broker', {}) if config else {},
                'device_count': 0
            })
    except Exception as e:
        current_app.logger.error(f"Error getting MQTT status: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500


@mqtt_bp.route('/config', methods=['GET'])
def get_mqtt_config():
    """Get MQTT configuration"""
    try:
        config = load_mqtt_config()
        if config:
            # Remove password from response for security
            if 'broker' in config and 'password' in config['broker']:
                config_copy = json.loads(json.dumps(config))  # Deep copy
                config_copy['broker']['password'] = '***' if config['broker']['password'] else ''
                return jsonify({'success': True, 'config': config_copy})
            return jsonify({'success': True, 'config': config})
        return jsonify({'success': False, 'error': 'Configuration not found'}), 404
    except Exception as e:
        current_app.logger.error(f"Error getting MQTT config: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500


@mqtt_bp.route('/config', methods=['POST'])
def update_mqtt_config():
    """Update MQTT configuration"""
    try:
        new_config = request.get_json()
        
        if not new_config:
            return jsonify({'success': False, 'error': 'No configuration provided'}), 400
        
        # Load existing config to preserve password if not provided
        existing_config = load_mqtt_config()
        if existing_config and 'broker' in new_config:
            if new_config['broker'].get('password') == '***':
                new_config['broker']['password'] = existing_config['broker'].get('password', '')
        
        # Save configuration
        if save_mqtt_config(new_config):
            current_app.logger.info("MQTT configuration updated")
            return jsonify({'success': True, 'message': 'Configuration updated'})
        
        return jsonify({'success': False, 'error': 'Failed to save configuration'}), 500
    except Exception as e:
        current_app.logger.error(f"Error updating MQTT config: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500


@mqtt_bp.route('/connect', methods=['POST'])
def mqtt_connect():
    """Connect to MQTT broker"""
    try:
        config = load_mqtt_config()
        if not config:
            return jsonify({'success': False, 'error': 'Configuration not found'}), 404
        
        broker = config.get('broker', {})
        
        # Initialize and connect client
        client = init_mqtt_client(
            broker_host=broker.get('host', 'localhost'),
            broker_port=broker.get('port', 1883),
            username=broker.get('username') or None,
            password=broker.get('password') or None
        )
        
        if client.connect():
            # Add configured devices
            for device in config.get('devices', []):
                client.add_device(
                    name=device['name'],
                    topic=device['topic'],
                    device_type=device.get('device_type', 'generic')
                )
            
            current_app.logger.info("Connected to MQTT broker")
            return jsonify({'success': True, 'message': 'Connected to MQTT broker'})
        
        return jsonify({'success': False, 'error': 'Failed to connect to broker'}), 500
    except Exception as e:
        current_app.logger.error(f"Error connecting to MQTT broker: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500


@mqtt_bp.route('/disconnect', methods=['POST'])
def mqtt_disconnect():
    """Disconnect from MQTT broker"""
    try:
        shutdown_mqtt_client()
        current_app.logger.info("Disconnected from MQTT broker")
        return jsonify({'success': True, 'message': 'Disconnected from MQTT broker'})
    except Exception as e:
        current_app.logger.error(f"Error disconnecting from MQTT broker: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500


@mqtt_bp.route('/devices', methods=['GET'])
def get_devices():
    """Get list of all MQTT devices"""
    try:
        client = get_mqtt_client()
        
        if not client:
            return jsonify({'success': False, 'error': 'MQTT client not initialized'}), 400
        
        devices = client.get_devices()
        return jsonify({'success': True, 'devices': devices})
    except Exception as e:
        current_app.logger.error(f"Error getting devices: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500


@mqtt_bp.route('/devices/<device_topic>', methods=['GET'])
def get_device(device_topic):
    """Get specific device status"""
    try:
        client = get_mqtt_client()
        
        if not client:
            return jsonify({'success': False, 'error': 'MQTT client not initialized'}), 400
        
        device = client.get_device(device_topic)
        
        if device:
            return jsonify({'success': True, 'device': device})
        
        return jsonify({'success': False, 'error': 'Device not found'}), 404
    except Exception as e:
        current_app.logger.error(f"Error getting device: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500


@mqtt_bp.route('/devices', methods=['POST'])
def add_device():
    """Add a new device"""
    try:
        data = request.get_json()
        
        if not data or not data.get('name') or not data.get('topic'):
            return jsonify({'success': False, 'error': 'Name and topic required'}), 400
        
        client = get_mqtt_client()
        if not client:
            return jsonify({'success': False, 'error': 'MQTT client not initialized'}), 400
        
        # Add device to client
        device = client.add_device(
            name=data['name'],
            topic=data['topic'],
            device_type=data.get('device_type', 'generic')
        )
        
        # Update configuration file
        config = load_mqtt_config()
        if config:
            config['devices'].append({
                'name': data['name'],
                'topic': data['topic'],
                'device_type': data.get('device_type', 'generic'),
                'description': data.get('description', '')
            })
            save_mqtt_config(config)
        
        current_app.logger.info(f"Added device: {data['name']}")
        return jsonify({'success': True, 'device': device.to_dict()})
    except Exception as e:
        current_app.logger.error(f"Error adding device: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500


@mqtt_bp.route('/devices/<device_topic>', methods=['DELETE'])
def remove_device(device_topic):
    """Remove a device"""
    try:
        client = get_mqtt_client()
        
        if not client:
            return jsonify({'success': False, 'error': 'MQTT client not initialized'}), 400
        
        # Remove from client
        if client.remove_device(device_topic):
            # Update configuration file
            config = load_mqtt_config()
            if config:
                config['devices'] = [d for d in config['devices'] if d['topic'] != device_topic]
                save_mqtt_config(config)
            
            current_app.logger.info(f"Removed device: {device_topic}")
            return jsonify({'success': True, 'message': 'Device removed'})
        
        return jsonify({'success': False, 'error': 'Device not found'}), 404
    except Exception as e:
        current_app.logger.error(f"Error removing device: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500


@mqtt_bp.route('/command', methods=['POST'])
def send_command():
    """Send command to a device"""
    try:
        data = request.get_json()
        
        if not data or not data.get('device_topic') or not data.get('command'):
            return jsonify({'success': False, 'error': 'Device topic and command required'}), 400
        
        client = get_mqtt_client()
        
        if not client or not client.is_connected():
            return jsonify({'success': False, 'error': 'MQTT client not connected'}), 400
        
        result = client.publish_command(
            device_topic=data['device_topic'],
            command=data['command'],
            payload=data.get('payload', '')
        )
        
        if result:
            current_app.logger.info(f"Sent command to {data['device_topic']}: {data['command']}")
            return jsonify({'success': True, 'message': 'Command sent'})
        
        return jsonify({'success': False, 'error': 'Failed to send command'}), 500
    except Exception as e:
        current_app.logger.error(f"Error sending command: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500


@mqtt_bp.route('/power/<device_topic>', methods=['POST'])
def control_power(device_topic):
    """Control device power (relay)"""
    try:
        data = request.get_json()
        
        if not data or 'action' not in data:
            return jsonify({'success': False, 'error': 'Action required (on/off/toggle)'}), 400
        
        client = get_mqtt_client()
        
        if not client or not client.is_connected():
            return jsonify({'success': False, 'error': 'MQTT client not connected'}), 400
        
        action = data['action'].lower()
        relay = data.get('relay')  # Optional relay number
        
        if action == 'on':
            result = client.power_on(device_topic, relay)
        elif action == 'off':
            result = client.power_off(device_topic, relay)
        elif action == 'toggle':
            result = client.power_toggle(device_topic, relay)
        else:
            return jsonify({'success': False, 'error': 'Invalid action'}), 400
        
        if result:
            current_app.logger.info(f"Power {action} for {device_topic}")
            return jsonify({'success': True, 'message': f'Power {action}'})
        
        return jsonify({'success': False, 'error': 'Failed to control power'}), 500
    except Exception as e:
        current_app.logger.error(f"Error controlling power: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500


@mqtt_bp.route('/status/<device_topic>', methods=['POST'])
def request_status(device_topic):
    """Request status update from device"""
    try:
        client = get_mqtt_client()
        
        if not client or not client.is_connected():
            return jsonify({'success': False, 'error': 'MQTT client not connected'}), 400
        
        data = request.get_json() or {}
        status_type = data.get('type', 0)
        
        result = client.get_status(device_topic, status_type)
        
        if result:
            return jsonify({'success': True, 'message': 'Status request sent'})
        
        return jsonify({'success': False, 'error': 'Failed to request status'}), 500
    except Exception as e:
        current_app.logger.error(f"Error requesting status: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500


@mqtt_bp.route('/dimmer/<device_topic>', methods=['POST'])
def control_dimmer(device_topic):
    """Control dimmer level"""
    try:
        data = request.get_json()
        
        if not data or 'level' not in data:
            return jsonify({'success': False, 'error': 'Level required (0-100)'}), 400
        
        client = get_mqtt_client()
        
        if not client or not client.is_connected():
            return jsonify({'success': False, 'error': 'MQTT client not connected'}), 400
        
        level = int(data['level'])
        if level < 0 or level > 100:
            return jsonify({'success': False, 'error': 'Level must be between 0 and 100'}), 400
        
        result = client.set_dimmer(device_topic, level)
        
        if result:
            return jsonify({'success': True, 'message': f'Dimmer set to {level}%'})
        
        return jsonify({'success': False, 'error': 'Failed to set dimmer'}), 500
    except Exception as e:
        current_app.logger.error(f"Error controlling dimmer: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500


@mqtt_bp.route('/color/<device_topic>', methods=['POST'])
def control_color(device_topic):
    """Control RGB color"""
    try:
        data = request.get_json()
        
        if not data or 'color' not in data:
            return jsonify({'success': False, 'error': 'Color required (hex format: RRGGBB)'}), 400
        
        client = get_mqtt_client()
        
        if not client or not client.is_connected():
            return jsonify({'success': False, 'error': 'MQTT client not connected'}), 400
        
        color = data['color'].replace('#', '')
        result = client.set_color(device_topic, color)
        
        if result:
            return jsonify({'success': True, 'message': f'Color set to #{color}'})
        
        return jsonify({'success': False, 'error': 'Failed to set color'}), 500
    except Exception as e:
        current_app.logger.error(f"Error controlling color: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500


@mqtt_bp.route('/restart/<device_topic>', methods=['POST'])
def restart_device(device_topic):
    """Restart a device"""
    try:
        client = get_mqtt_client()
        
        if not client or not client.is_connected():
            return jsonify({'success': False, 'error': 'MQTT client not connected'}), 400
        
        result = client.restart(device_topic)
        
        if result:
            return jsonify({'success': True, 'message': 'Restart command sent'})
        
        return jsonify({'success': False, 'error': 'Failed to restart device'}), 500
    except Exception as e:
        current_app.logger.error(f"Error restarting device: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500

File: app/static/js/mqtt.js
/**
 * MQTT/IoT Device Control JavaScript
 * Handles Tasmota ESP32 device control via MQTT
 */

// Global state
let mqttConnected = false;
let devices = [];
let currentDeviceTopic = null;

// Initialize on page load
document.addEventListener('DOMContentLoaded', function() {
  loadMQTTStatus();
  
  // Set up event listeners
  document.getElementById('connect-btn').addEventListener('click', connectToMQTT);
  document.getElementById('disconnect-btn').addEventListener('click', disconnectFromMQTT);
  document.getElementById('config-btn').addEventListener('click', openConfigModal);
  document.getElementById('add-device-btn').addEventListener('click', openAddDeviceModal);
  document.getElementById('mqtt-config-form').addEventListener('submit', saveConfiguration);
  document.getElementById('add-device-form').addEventListener('submit', addDevice);
  
  // Auto-refresh status every 5 seconds
  setInterval(loadMQTTStatus, 5000);
});

/**
 * Show toast notification
 */
function showToast(message, type = 'info') {
  const container = document.getElementById('toast-container');
  const toast = document.createElement('div');
  toast.className = `toast ${type}`;
  toast.textContent = message;
  
  container.appendChild(toast);
  
  // Remove after 4 seconds
  setTimeout(() => {
    toast.style.animation = 'slideIn 0.3s ease reverse';
    setTimeout(() => toast.remove(), 300);
  }, 4000);
}

/**
 * Load MQTT connection status
 */
async function loadMQTTStatus() {
  try {
    const response = await fetch('/api/mqtt/status');
    const data = await response.json();
    
    if (data.success) {
      mqttConnected = data.connected;
      updateStatusUI(data);
      
      if (mqttConnected) {
        loadDevices();
      }
    }
  } catch (error) {
    console.error('Error loading MQTT status:', error);
  }
}

/**
 * Update status UI
 */
function updateStatusUI(status) {
  const indicator = document.getElementById('mqtt-status-indicator');
  const statusText = document.getElementById('mqtt-status-text');
  const brokerInfo = document.getElementById('mqtt-broker-info');
  const connectBtn = document.getElementById('connect-btn');
  const disconnectBtn = document.getElementById('disconnect-btn');
  
  if (status.connected) {
    indicator.textContent = '🟢';
    statusText.textContent = 'Connected';
    brokerInfo.textContent = `${status.broker.host}:${status.broker.port} • ${status.device_count} device(s)`;
    connectBtn.disabled = true;
    disconnectBtn.disabled = false;
  } else {
    indicator.textContent = '🔴';
    statusText.textContent = 'Disconnected';
    
    if (status.broker && status.broker.host) {
      brokerInfo.textContent = `${status.broker.host}:${status.broker.port}`;
    } else {
      brokerInfo.textContent = 'Not configured';
    }
    
    connectBtn.disabled = false;
    disconnectBtn.disabled = true;
  }
}

/**
 * Connect to MQTT broker
 */
async function connectToMQTT() {
  try {
    showToast('Connecting to MQTT broker...', 'info');
    
    const response = await fetch('/api/mqtt/connect', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' }
    });
    
    const data = await response.json();
    
    if (data.success) {
      showToast('Connected to MQTT broker!', 'success');
      loadMQTTStatus();
    } else {
      showToast(`Connection failed: ${data.error}`, 'error');
    }
  } catch (error) {
    console.error('Error connecting to MQTT:', error);
    showToast('Failed to connect to MQTT broker', 'error');
  }
}

/**
 * Disconnect from MQTT broker
 */
async function disconnectFromMQTT() {
  try {
    const response = await fetch('/api/mqtt/disconnect', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' }
    });
    
    const data = await response.json();
    
    if (data.success) {
      showToast('Disconnected from MQTT broker', 'info');
      loadMQTTStatus();
      
      // Clear devices display
      const grid = document.getElementById('devices-grid');
      grid.innerHTML = '<div class="loading-message">Connect to MQTT broker to view devices...</div>';
    } else {
      showToast(`Disconnect failed: ${data.error}`, 'error');
    }
  } catch (error) {
    console.error('Error disconnecting from MQTT:', error);
    showToast('Failed to disconnect from MQTT broker', 'error');
  }
}

/**
 * Load devices
 */
async function loadDevices() {
  try {
    const response = await fetch('/api/mqtt/devices');
    const data = await response.json();
    
    if (data.success) {
      devices = data.devices;
      renderDevices(devices);
    }
  } catch (error) {
    console.error('Error loading devices:', error);
  }
}

/**
 * Render devices grid
 */
function renderDevices(devices) {
  const grid = document.getElementById('devices-grid');
  
  if (devices.length === 0) {
    grid.innerHTML = '<div class="loading-message">No devices configured. Click "Add Device" to get started.</div>';
    return;
  }
  
  grid.innerHTML = '';
  
  devices.forEach(device => {
    const card = createDeviceCard(device);
    grid.appendChild(card);
  });
}

/**
 * Create device card element
 */
function createDeviceCard(device) {
  const card = document.createElement('div');
  card.className = `device-card ${device.online ? '' : 'offline'}`;
  
  const statusClass = device.online ? 'online' : 'offline';
  const statusText = device.online ? 'Online' : 'Offline';
  
  // Get device type icon
  const typeIcons = {
    'switch': '🔌',
    'light': '💡',
    'rgb': '🌈',
    'sensor': '📊',
    'generic': '📡'
  };
  const typeIcon = typeIcons[device.device_type] || '📡';
  
  card.innerHTML = `
    <div class="device-header">
      <div>
        <div class="device-name">${typeIcon} ${device.name}</div>
        <div class="device-topic">${device.topic}</div>
        <span class="device-type-badge">${device.device_type}</span>
      </div>
      <span class="device-status ${statusClass}">${statusText}</span>
    </div>
    <div class="device-info">
      ${device.last_seen ? `Last seen: ${new Date(device.last_seen * 1000).toLocaleString()}` : 'Never seen'}
    </div>
    <div class="device-actions">
      <button class="btn btn-success btn-small" onclick="controlPower('${device.topic}', 'on')" ${!device.online ? 'disabled' : ''}>
        ON
      </button>
      <button class="btn btn-danger btn-small" onclick="controlPower('${device.topic}', 'off')" ${!device.online ? 'disabled' : ''}>
        OFF
      </button>
      <button class="btn btn-warning btn-small" onclick="controlPower('${device.topic}', 'toggle')" ${!device.online ? 'disabled' : ''}>
        TOGGLE
      </button>
      <button class="btn btn-secondary btn-small" onclick="openControlModal('${device.topic}')" ${!device.online ? 'disabled' : ''}>
        More...
      </button>
      <button class="btn btn-secondary btn-small" onclick="removeDevice('${device.topic}')">
        🗑️
      </button>
    </div>
  `;
  
  return card;
}

/**
 * Control device power
 */
async function controlPower(deviceTopic, action) {
  try {
    const response = await fetch(`/api/mqtt/power/${deviceTopic}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ action })
    });
    
    const data = await response.json();
    
    if (data.success) {
      showToast(`Power ${action} command sent`, 'success');
      // Refresh devices after a short delay
      setTimeout(loadDevices, 1000);
    } else {
      showToast(`Failed: ${data.error}`, 'error');
    }
  } catch (error) {
    console.error('Error controlling power:', error);
    showToast('Failed to send command', 'error');
  }
}

/**
 * Open configuration modal
 */
async function openConfigModal() {
  try {
    const response = await fetch('/api/mqtt/config');
    const data = await response.json();
    
    if (data.success) {
      const config = data.config;
      
      document.getElementById('broker-host').value = config.broker.host || 'localhost';
      document.getElementById('broker-port').value = config.broker.port || 1883;
      document.getElementById('broker-username').value = config.broker.username || '';
      document.getElementById('broker-password').value = config.broker.password || '';
      
      document.getElementById('config-modal').style.display = 'flex';
    }
  } catch (error) {
    console.error('Error loading config:', error);
    showToast('Failed to load configuration', 'error');
  }
}

/**
 * Close configuration modal
 */
function closeConfigModal() {
  document.getElementById('config-modal').style.display = 'none';
}

/**
 * Save configuration
 */
async function saveConfiguration(e) {
  e.preventDefault();
  
  const config = {
    broker: {
      host: document.getElementById('broker-host').value,
      port: parseInt(document.getElementById('broker-port').value),
      username: document.getElementById('broker-username').value,
      password: document.getElementById('broker-password').value,
      auto_connect: true
    },
    devices: devices.map(d => ({
      name: d.name,
      topic: d.topic,
      device_type: d.device_type,
      description: ''
    })),
    settings: {
      discovery_enabled: true,
      retain_messages: false,
      qos: 1
    }
  };
  
  try {
    const response = await fetch('/api/mqtt/config', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(config)
    });
    
    const data = await response.json();
    
    if (data.success) {
      showToast('Configuration saved successfully', 'success');
      closeConfigModal();
      loadMQTTStatus();
    } else {
      showToast(`Failed to save: ${data.error}`, 'error');
    }
  } catch (error) {
    console.error('Error saving config:', error);
    showToast('Failed to save configuration', 'error');
  }
}

/**
 * Open add device modal
 */
function openAddDeviceModal() {
  document.getElementById('add-device-form').reset();
  document.getElementById('add-device-modal').style.display = 'flex';
}

/**
 * Close add device modal
 */
function closeAddDeviceModal() {
  document.getElementById('add-device-modal').style.display = 'none';
}

/**
 * Add new device
 */
async function addDevice(e) {
  e.preventDefault();
  
  const deviceData = {
    name: document.getElementById('device-name').value,
    topic: document.getElementById('device-topic').value,
    device_type: document.getElementById('device-type').value,
    description: document.getElementById('device-description').value
  };
  
  try {
    const response = await fetch('/api/mqtt/devices', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(deviceData)
    });
    
    const data = await response.json();
    
    if (data.success) {
      showToast(`Device "${deviceData.name}" added successfully`, 'success');
      closeAddDeviceModal();
      loadDevices();
    } else {
      showToast(`Failed to add device: ${data.error}`, 'error');
    }
  } catch (error) {
    console.error('Error adding device:', error);
    showToast('Failed to add device', 'error');
  }
}

/**
 * Remove device
 */
async function removeDevice(deviceTopic) {
  if (!confirm('Are you sure you want to remove this device?')) {
    return;
  }
  
  try {
    const response = await fetch(`/api/mqtt/devices/${deviceTopic}`, {
      method: 'DELETE'
    });
    
    const data = await response.json();
    
    if (data.success) {
      showToast('Device removed successfully', 'success');
      loadDevices();
    } else {
      showToast(`Failed to remove device: ${data.error}`, 'error');
    }
  } catch (error) {
    console.error('Error removing device:', error);
    showToast('Failed to remove device', 'error');
  }
}

/**
 * Open device control modal
 */
function openControlModal(deviceTopic) {
  currentDeviceTopic = deviceTopic;
  const device = devices.find(d => d.topic === deviceTopic);
  
  if (!device) return;
  
  document.getElementById('control-device-name').textContent = device.name;
  
  // Generate control options based on device type
  let controlHTML = '';
  
  if (device.device_type === 'light' || device.device_type === 'rgb') {
    controlHTML = `
      <div class="form-group">
        <label>Dimmer Level (0-100%):</label>
        <input type="range" id="dimmer-slider" min="0" max="100" value="50" style="width: 100%;">
        <div style="text-align: center; margin-top: 0.5rem;">
          <span id="dimmer-value">50</span>%
        </div>
        <button class="btn btn-primary" onclick="setDimmer()" style="width: 100%; margin-top: 0.5rem;">Set Dimmer</button>
      </div>
    `;
  }
  
  if (device.device_type === 'rgb') {
    controlHTML += `
      <div class="form-group" style="margin-top: 1rem;">
        <label>RGB Color:</label>
        <input type="color" id="color-picker" value="#ffffff" style="width: 100%; height: 50px; cursor: pointer;">
        <button class="btn btn-primary" onclick="setColor()" style="width: 100%; margin-top: 0.5rem;">Set Color</button>
      </div>
    `;
  }
  
  controlHTML += `
    <div class="form-group" style="margin-top: 1rem;">
      <label>Advanced Commands:</label>
      <button class="btn btn-secondary" onclick="requestStatus()" style="width: 100%; margin-bottom: 0.5rem;">Request Status</button>
      <button class="btn btn-warning" onclick="restartDevice()" style="width: 100%;">Restart Device</button>
    </div>
  `;
  
  document.getElementById('control-content').innerHTML = controlHTML;
  
  // Add event listener for dimmer slider
  const dimmerSlider = document.getElementById('dimmer-slider');
  if (dimmerSlider) {
    dimmerSlider.addEventListener('input', function() {
      document.getElementById('dimmer-value').textContent = this.value;
    });
  }
  
  document.getElementById('device-control-modal').style.display = 'flex';
}

/**
 * Close control modal
 */
function closeControlModal() {
  document.getElementById('device-control-modal').style.display = 'none';
  currentDeviceTopic = null;
}

/**
 * Set dimmer level
 */
async function setDimmer() {
  const level = document.getElementById('dimmer-slider').value;
  
  try {
    const response = await fetch(`/api/mqtt/dimmer/${currentDeviceTopic}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ level: parseInt(level) })
    });
    
    const data = await response.json();
    
    if (data.success) {
      showToast(`Dimmer set to ${level}%`, 'success');
    } else {
      showToast(`Failed: ${data.error}`, 'error');
    }
  } catch (error) {
    console.error('Error setting dimmer:', error);
    showToast('Failed to set dimmer', 'error');
  }
}

/**
 * Set RGB color
 */
async function setColor() {
  const color = document.getElementById('color-picker').value;
  
  try {
    const response = await fetch(`/api/mqtt/color/${currentDeviceTopic}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ color })
    });
    
    const data = await response.json();
    
    if (data.success) {
      showToast(`Color set to ${color}`, 'success');
    } else {
      showToast(`Failed: ${data.error}`, 'error');
    }
  } catch (error) {
    console.error('Error setting color:', error);
    showToast('Failed to set color', 'error');
  }
}

/**
 * Request device status
 */
async function requestStatus() {
  try {
    const response = await fetch(`/api/mqtt/status/${currentDeviceTopic}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ type: 0 })
    });
    
    const data = await response.json();
    
    if (data.success) {
      showToast('Status request sent', 'info');
    } else {
      showToast(`Failed: ${data.error}`, 'error');
    }
  } catch (error) {
    console.error('Error requesting status:', error);
    showToast('Failed to request status', 'error');
  }
}

/**
 * Restart device
 */
async function restartDevice() {
  if (!confirm('Are you sure you want to restart this device?')) {
    return;
  }
  
  try {
    const response = await fetch(`/api/mqtt/restart/${currentDeviceTopic}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' }
    });
    
    const data = await response.json();
    
    if (data.success) {
      showToast('Restart command sent', 'warning');
      closeControlModal();
    } else {
      showToast(`Failed: ${data.error}`, 'error');
    }
  } catch (error) {
    console.error('Error restarting device:', error);
    showToast('Failed to restart device', 'error');
  }
}

// Close modals when clicking outside
window.onclick = function(event) {
  const configModal = document.getElementById('config-modal');
  const addDeviceModal = document.getElementById('add-device-modal');
  const controlModal = document.getElementById('device-control-modal');
  
  if (event.target === configModal) {
    closeConfigModal();
  } else if (event.target === addDeviceModal) {
    closeAddDeviceModal();
  } else if (event.target === controlModal) {
    closeControlModal();
  }
};

